lib/usdUfe/ufe/UsdUndoAddRefOrPayloadCommand.cpp

```diff
@@ -0,0 +1,141 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdUndoAddRefOrPayloadCommand.h"
+
+#include <pxr/base/tf/stringUtils.h>
+#include <pxr/usd/sdf/layer.h>
+#include <pxr/usd/sdf/path.h>
+#include <pxr/usd/usd/payloads.h>
+#include <pxr/usd/usd/references.h>
+
+namespace USDUFE_NS_DEF {
+
+using namespace PXR_NS;
+
+static std::string validatePrimSpec(const UsdPrim& prim, const SdfPrimSpecHandle& primSpec)
+{
+    if (!primSpec)
+        return "is not valid.";
+
+    // A common error is to reference a prim that is not the same type as the prim
+    // that contains the reference. Since only the type of the prim that contains
+    // the reference is used, the referenced prim might not show up.
+    //
+    // This happens a lot when trying to reference geometry (mesh) instead of the
+    // prim containing the geometry. Of vis-versa, referencing a prim inside a mesh.
+    const std::string& primType = prim.GetTypeName();
+    const std::string& targetType = primSpec->GetTypeName();
+    if (primType != targetType)
+        return TfStringPrintf(
+            "does not have the same type as the targeted prim: [%s] vs [%s].",
+            primType.c_str(),
+            targetType.c_str());
+
+    return "";
+}
+
+static PXR_NS::SdfPath
+getPrimPath(const UsdPrim& prim, const std::string& filePath, const std::string& primPath)
+{
+    // When an explicit prim path was given, use that.
+    if (!primPath.empty())
+        return SdfPath(primPath);
+
+    // If no prim path were specified and we are referencing a MaterialX file
+    // then use the MaterialX prim as the target for the reference.
+    //
+    // TODO: should we force this even when the referenced file has a default prim?
+    if (TfStringEndsWith(filePath, ".mtlx"))
+        return SdfPath("/MaterialX");
+
+    // Retrieve the layer for analysis.
+    //
+    // Note: we don't print any warning if the layer cannot be found as we assume
+    //       the load itself will also fail and print an error.
+    SdfLayerRefPtr layerRef = SdfLayer::FindOrOpen(filePath);
+    if (!layerRef)
+        return SdfPath();
+
+    // If the referenced file has a default prim, leave the prim path empty.
+    if (layerRef->HasDefaultPrim()) {
+        TfToken           primName = layerRef->GetDefaultPrim();
+        SdfPrimSpecHandle primSpec = layerRef->GetPrimAtPath(SdfPath(primName.GetText()));
+        const std::string errorMessage = validatePrimSpec(prim, primSpec);
+        if (!errorMessage.empty())
+            TF_WARN(
+                (std::string("The default prim in file [%s] ") + errorMessage).c_str(),
+                filePath.c_str());
+        return SdfPath();
+    }
+
+    // If the referenced file has no default prim, return the path to the first
+    // valid root prim we find.
+    std::string errorMessage = " is absent.";
```

Comment:
The errorMessage is the rror about the prim, the gerenal message is about the referenced file. For example, bu default it reads: "The file [a/b/c] does not contain a default prim and the root prim is absent."

In othe rcase it might read: "The file [a/b/c] does not contain a default prim and the root prim is invalid." or " does not have the same type as the referencing prim."

Reply:
I would suggest to output the first warning any time the file doesn't have a default prim and the second when we fail to find an acceptable root prim.

Reply:
OK, I could split this up. The goal is that the validation of the default prim (when there is one) or the chosen root prim use the same validation code which generates the error message., see line 79